# YouTube Trending Video EDA CSV

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("../../data/trends/USvideos.csv")
df.head(5)

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   

   category_id              publish_time  \
0           22  2017-11-13T17:13:01.000Z   
1           24  2017-11-13T07:30:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-13T11:00:04.000Z   
4           24  2017-11-12T18:01:41.000Z   

                                                tags    views   likes  \
0                                    SHANtell martin   748374   57527   
1  last week tonight trump presidency|"last week ...  2418783   97185   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434  146033   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168   10172   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731  132235   

   dislikes  comment_count                                  thumbnail_link  \
0      2966          15954  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1      6146          12703  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3       666           2146  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  
1  One year after the presidential election, John...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  Today we find out if Link is a Nickelback amat...  
4  I know it's been a while since we did this sho...

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40949 entries, 0 to 40948
Data columns (total 16 columns):
video_id                  40949 non-null object
trending_date             40949 non-null object
title                     40949 non-null object
channel_title             40949 non-null object
category_id               40949 non-null int64
publish_time              40949 non-null object
tags                      40949 non-null object
views                     40949 non-null int64
likes                     40949 non-null int64
dislikes                  40949 non-null int64
comment_count             40949 non-null int64
thumbnail_link            40949 non-null object
comments_disabled         40949 non-null bool
ratings_disabled          40949 non-null bool
video_error_or_removed    40949 non-null bool
description               40379 non-null object
dtypes: bool(3), int64(5), object(8)
memory usage: 4.2+ MB


In [4]:
df[df["description"].apply(lambda x: pd.isna(x))].head(3)

video_id trending_date  \
42   NZFhMSgbKKM      17.14.11   
47   sbcbvuitiTc      17.14.11   
175  4d07RXYLsJE      17.14.11   

                                                 title  \
42   Dennis Smith Jr. and LeBron James go back and ...   
47   Stephon Marbury and Jimmer Fredette fight in C...   
175                Sphaera - demonstrating interaction   

                channel_title  category_id              publish_time  \
42               Ben Rohrbach           17  2017-11-13T15:11:00.000Z   
47   NBA Highlights · YouTube           17  2017-11-10T18:23:05.000Z   
175              Jenny Hanell           28  2017-11-04T20:48:16.000Z   

                          tags   views  likes  dislikes  comment_count  \
42                      [none]     945      7         5              8   
47   NBA|"Basketball"|"Sports"  956169   2017      2425           1447   
175                     [none]    1827      3         0              2   

                                     thumbnail_link  comments_disabled  \
42   https://i.ytimg.com/vi/NZFhMSgbKKM/default.jpg              False   
47   https://i.ytimg.com/vi/sbcbvuitiTc/default.jpg              False   
175  https://i.ytimg.com/vi/4d07RXYLsJE/default.jpg              False   

     ratings_disabled  video_error_or_removed description  
42              False                   False         NaN  
47              False                   False         NaN  
175             False                   False         NaN

In [5]:
df["description"] = df["description"].fillna(value="")

In [6]:
df["trending_date"].sort_values().head(3)

3400    17.01.12
3528    17.01.12
3529    17.01.12
Name: trending_date, dtype: object

In [7]:
df["trending_date"].sort_values(ascending=False).head(3)

38144    18.31.05
38010    18.31.05
38020    18.31.05
Name: trending_date, dtype: object

Sorting "trending_date" with sort_values is not giving earliest and latest days as desired because format is year.day.month. Was able to use output from df["trending_date"].unique() to find earliest and latest days

In [8]:
df["trending_date"].unique().size

205

# '17.14.11' is earliest day. '18.14.06' is latest day.  
11/14/2017 --- 6/14/2018  Using time calculator: 213 days (7months)  
213-205=8 This means that 8 days are missing trending video data. (Given "trending_date" size above)

In [9]:
df["video_id"].unique().size

6351

# Videos appear multiple times. 6351 videos. 40949 rows

In [10]:
df.describe()

category_id         views         likes      dislikes  comment_count
count  40949.000000  4.094900e+04  4.094900e+04  4.094900e+04   4.094900e+04
mean      19.972429  2.360785e+06  7.426670e+04  3.711401e+03   8.446804e+03
std        7.568327  7.394114e+06  2.288853e+05  2.902971e+04   3.743049e+04
min        1.000000  5.490000e+02  0.000000e+00  0.000000e+00   0.000000e+00
25%       17.000000  2.423290e+05  5.424000e+03  2.020000e+02   6.140000e+02
50%       24.000000  6.818610e+05  1.809100e+04  6.310000e+02   1.856000e+03
75%       25.000000  1.823157e+06  5.541700e+04  1.938000e+03   5.755000e+03
max       43.000000  2.252119e+08  5.613827e+06  1.674420e+06   1.361580e+06

In [11]:
df["views"].sort_values(ascending=False).head()

38547    225211923
38345    220490543
38146    217750076
37935    210338856
37730    205643016
Name: views, dtype: int64

# View count range suprisingly large 549 - 225,211,923

In [12]:
df.describe(include = ['O'])

video_id trending_date                                       title  \
count         40949         40949                                       40949   
unique         6351           205                                        6455   
top     j4KvrAUjn6c      17.03.12  WE MADE OUR MOM CRY...HER DREAM CAME TRUE!   
freq             30           200                                          30   

       channel_title              publish_time    tags  \
count          40949                     40949   40949   
unique          2207                      6269    6055   
top             ESPN  2018-05-18T14:00:04.000Z  [none]   
freq             203                        50    1535   

                                        thumbnail_link description  
count                                            40949       40949  
unique                                            6352        6902  
top     https://i.ytimg.com/vi/j4KvrAUjn6c/default.jpg              
freq                                                30         570

# 6055 / 6351 unique tags to unique videos  
# The tag data has good variety and usefulness

In [13]:
df["tags"].unique()[25:30]

array(['spaghetti burrito|"diy burrito"|"spaghetti"|"burrito"|"burrito recipe"|"spaghetti deep fried"|"deep fried spaghetti"|"spgahetti sandwich"|"giant burrito"',
       'numberphile|"prime numbers"|"proth prime"',
       'Smart mug|"Heated thermos"|"tech"|"gift idea"|"unboxed"|"Ember mug"|"ember"|"Ember review"|"Heated mug"|"teardown"|"technology"|"Thermous"|"winter"|"holiday"|"christmas"|"gift"|"mug"|"unboxing"',
       'auth-jvardon-auth',
       'freaks and geeks|"jason segel"|"judd apatow"|"drums"|"rush"|"paul feig"|"drummer"|"nick andopolis"'],
      dtype=object)

In [14]:
df.corr()

category_id     views     likes  dislikes  \
category_id                1.000000 -0.168231 -0.173921 -0.033547   
views                     -0.168231  1.000000  0.849177  0.472213   
likes                     -0.173921  0.849177  1.000000  0.447186   
dislikes                  -0.033547  0.472213  0.447186  1.000000   
comment_count             -0.076307  0.617621  0.803057  0.700184   
comments_disabled          0.048949  0.002677 -0.028918 -0.004431   
ratings_disabled          -0.013506  0.015355 -0.020888 -0.008230   
video_error_or_removed    -0.030011 -0.002256 -0.002641 -0.001853   

                        comment_count  comments_disabled  ratings_disabled  \
category_id                 -0.076307           0.048949         -0.013506   
views                        0.617621           0.002677          0.015355   
likes                        0.803057          -0.028918         -0.020888   
dislikes                     0.700184          -0.004431         -0.008230   
comment_count                1.000000          -0.028277         -0.013819   
comments_disabled           -0.028277           1.000000          0.319230   
ratings_disabled            -0.013819           0.319230          1.000000   
video_error_or_removed      -0.003725          -0.002970         -0.001526   

                        video_error_or_removed  
category_id                          -0.030011  
views                                -0.002256  
likes                                -0.002641  
dislikes                             -0.001853  
comment_count                        -0.003725  
comments_disabled                    -0.002970  
ratings_disabled                     -0.001526  
video_error_or_removed                1.000000

Viewcount is understandably corellated to likes (0.849), dislikes (0.472), and comment count (0.618)

In [15]:
np.sort(df["category_id"].unique())

array([ 1,  2, 10, 15, 17, 19, 20, 22, 23, 24, 25, 26, 27, 28, 29, 43])

In [16]:
# tags_list = df['tags'].str.replace('"', '').str.split('|')
tags_list = pd.Series(df['tags'].str.replace('"', '').str.split('|'))
# df['tags_count'] = pd.series(tags_list).value_counts
# df['tags'].str.replace('"', '').str.split('|')[10][-15:]
# pd.concat(tags_list)
tags_list.value_counts().head(3)

[[none]]                                                                                                                                                                                                                                                                                                                                                                                                                1535
[ABC, americanidol, idol, american idol, ryan, seacrest, ryan seacrest, katy, perry, katy perry, luke, bryan, luke bryan, lionel, richie, lionel richie, season 16, american idol XVI, television, ad, spring, 2018, music, reality, competition, song, sing, audition, auditions, performance, live, fox, AI, hollywood, contestant, official, american, official american idol, hollywood week, hometown audition]      87
[Jacksfilms, Jack Douglass, YGS, YGS 100, YGS 50, The Best of Your Grammar Sucks, Your Grammar Sucks, YIAY, Yesterday I Asked You, Fidget Spinners, Emoji Movie, Kermit Sings,

In [24]:
dfsel = df.groupby(['video_id']).max().copy()
dfsel.head(2)

trending_date                                              title  \
video_id                                                                       
-0CMnp02rNY      18.11.06  Mindy Kaling's Daughter Had the Perfect Reacti...   
-0NYY8cqdiQ      18.01.02  Megan Mullally Didn't Notice the Interesting P...   

            channel_title  category_id              publish_time  \
video_id                                                           
-0CMnp02rNY  TheEllenShow           24  2018-06-04T13:00:00.000Z   
-0NYY8cqdiQ  TheEllenShow           24  2018-01-29T14:00:39.000Z   

                                                          tags   views  likes  \
video_id                                                                        
-0CMnp02rNY  ellen|"ellen degeneres"|"the ellen show"|"elle...  800359   9773   
-0NYY8cqdiQ  megan mullally|"megan"|"mullally"|"will and gr...  563746   4429   

             dislikes  comment_count  \
video_id                               
-0CMnp02rNY       332            423   
-0NYY8cqdiQ        54             94   

                                             thumbnail_link  \
video_id                                                      
-0CMnp02rNY  https://i.ytimg.com/vi/-0CMnp02rNY/default.jpg   
-0NYY8cqdiQ  https://i.ytimg.com/vi/-0NYY8cqdiQ/default.jpg   

             comments_disabled  ratings_disabled  video_error_or_removed  \
video_id                                                                   
-0CMnp02rNY              False             False                   False   
-0NYY8cqdiQ              False             False                   False   

                                                   description  
video_id                                                        
-0CMnp02rNY  Ocean's 8 star Mindy Kaling dished on bringing...  
-0NYY8cqdiQ  Ellen and Megan Mullally have known each other...

In [18]:
tags_sel = pd.Series(dfsel['tags'].str.replace('"', '').str.split('|'))
flatten_matrix = [val for sublist in tags_sel for val in sublist]
# list_all = []
# [[list_all.append(tag) for tag in t_list] for t_list in tags_sel]
# len(list_all)
len(flatten_matrix)
# flatten_matrix[:9]

126729

In [19]:
pd.Series(flatten_matrix).unique().size

56506

In [20]:
# list_for_count = 
dfsel['tags_count'] = tags_sel.apply(lambda x: len(x))
# len(dfsel['tags'])
dfsel[['tags', 'tags_count']].head()

tags  tags_count
video_id                                                                  
-0CMnp02rNY  ellen|"ellen degeneres"|"the ellen show"|"elle...          33
-0NYY8cqdiQ  megan mullally|"megan"|"mullally"|"will and gr...          33
-1Hm41N0dUs  jimmy|"jimmy kimmel"|"jimmy kimmel live"|"late...          23
-1yT-K3c6YI  youtube quiz|"youtuber quiz"|"truth or dare"|"...           8
-2RVw2_QyxQ                         Chess|"Saint Louis"|"Club"           3

In [21]:
# dfsel[dfsel['tags'] == "[none]"].shape
# dfsel.loc[:, 'tags_count'][dfsel['tags'] == "[none]"] = 0
dfsel.loc[dfsel['tags'] == "[none]", ['tags_count']] = 0
# dfsel[['tags', 'tags_count']]

In [22]:
dfsel.head(3)

trending_date                                              title  \
video_id                                                                       
-0CMnp02rNY      18.11.06  Mindy Kaling's Daughter Had the Perfect Reacti...   
-0NYY8cqdiQ      18.01.02  Megan Mullally Didn't Notice the Interesting P...   
-1Hm41N0dUs      18.30.04  Cast of Avengers: Infinity War Draws Their Cha...   

                 channel_title  category_id              publish_time  \
video_id                                                                
-0CMnp02rNY       TheEllenShow           24  2018-06-04T13:00:00.000Z   
-0NYY8cqdiQ       TheEllenShow           24  2018-01-29T14:00:39.000Z   
-1Hm41N0dUs  Jimmy Kimmel Live           23  2018-04-27T07:30:02.000Z   

                                                          tags    views  \
video_id                                                                  
-0CMnp02rNY  ellen|"ellen degeneres"|"the ellen show"|"elle...   800359   
-0NYY8cqdiQ  megan mullally|"megan"|"mullally"|"will and gr...   563746   
-1Hm41N0dUs  jimmy|"jimmy kimmel"|"jimmy kimmel live"|"late...  2058516   

             likes  dislikes  comment_count  \
video_id                                      
-0CMnp02rNY   9773       332            423   
-0NYY8cqdiQ   4429        54             94   
-1Hm41N0dUs  41248       580           1490   

                                             thumbnail_link  \
video_id                                                      
-0CMnp02rNY  https://i.ytimg.com/vi/-0CMnp02rNY/default.jpg   
-0NYY8cqdiQ  https://i.ytimg.com/vi/-0NYY8cqdiQ/default.jpg   
-1Hm41N0dUs  https://i.ytimg.com/vi/-1Hm41N0dUs/default.jpg   

             comments_disabled  ratings_disabled  video_error_or_removed  \
video_id                                                                   
-0CMnp02rNY              False             False                   False   
-0NYY8cqdiQ              False             False                   False   
-1Hm41N0dUs              False             False                   False   

                                                   description  tags_count  
video_id                                                                    
-0CMnp02rNY  Ocean's 8 star Mindy Kaling dished on bringing...          33  
-0NYY8cqdiQ  Ellen and Megan Mullally have known each other...          33  
-1Hm41N0dUs  Benedict Cumberbatch, Don Cheadle, Elizabeth O...          23

In [53]:
# pd.Series(['1', '2', '3a', '3b', '03c']).str.match(pattern)
pat_sci_tech = 'science|technology'
pat_house = 'house|home|apartment|condo|townehome|ranch|castle|real estate|property'
pat_airbnb = 'airbnb|hotel|hostel|motel|rental|vacation|tourist|travel|destination|passenger'
pat_food = 'food|cook|kitchen|restaurant|breakfast|lunch|dinner|supper|hungry|tasty'
dfsel.loc[dfsel["tags"].str.contains(pat_sci_tech), ["tags"]].shape

(277, 1)

In [61]:
# dfsel.loc[dfsel["tags"].str.contains(pat_sci_tech), ["category_id"]]
df_sci_tech = dfsel.loc[dfsel["tags"].str.contains(pat_sci_tech)]
df_sci_tech.reset_index().groupby("category_id").count()

video_id  trending_date  title  channel_title  publish_time  \
category_id                                                                
1                   5              5      5              5             5   
2                   3              3      3              3             3   
10                  1              1      1              1             1   
15                  8              8      8              8             8   
17                  5              5      5              5             5   
19                  2              2      2              2             2   
22                 12             12     12             12            12   
23                  3              3      3              3             3   
24                 19             19     19             19            19   
25                 25             25     25             25            25   
26                  5              5      5              5             5   
27                 98             98     98             98            98   
28                 91             91     91             91            91   

             tags  views  likes  dislikes  comment_count  thumbnail_link  \
category_id                                                                
1               5      5      5         5              5               5   
2               3      3      3         3              3               3   
10              1      1      1         1              1               1   
15              8      8      8         8              8               8   
17              5      5      5         5              5               5   
19              2      2      2         2              2               2   
22             12     12     12        12             12              12   
23              3      3      3         3              3               3   
24             19     19     19        19             19              19   
25             25     25     25        25             25              25   
26              5      5      5         5              5               5   
27             98     98     98        98             98              98   
28             91     91     91        91             91              91   

             comments_disabled  ratings_disabled  video_error_or_removed  \
category_id                                                                
1                            5                 5                       5   
2                            3                 3                       3   
10                           1                 1                       1   
15                           8                 8                       8   
17                           5                 5                       5   
19                           2                 2                       2   
22                          12                12                      12   
23                           3                 3                       3   
24                          19                19                      19   
25                          25                25                      25   
26                           5                 5                       5   
27                          98                98                      98   
28                          91                91                      91   

             description  
category_id               
1                      5  
2                      3  
10                     1  
15                     8  
17                     5  
19                     2  
22                    12  
23                     3  
24                    19  
25                    25  
26                     5  
27                    98  
28                    91

In [62]:
dfsel.sort_values('views', ascending=False).head()

trending_date                                              title  \
video_id                                                                       
VYOjWnS4cMY      18.31.05  Childish Gambino - This Is America (Official V...   
FlsCjmMhFmw      17.14.12  YouTube Rewind: The Shape of 2017 | #YouTubeRe...   
ffxKSjUwKdU      18.30.04               Ariana Grande - No Tears Left To Cry   
zEf423kYfqk      18.30.04  Becky G, Natti Natasha - Sin Pijama (Official ...   
7C2z4GqqS5E      18.31.05                BTS (방탄소년단) 'FAKE LOVE' Official MV   

                   channel_title  category_id              publish_time  \
video_id                                                                  
VYOjWnS4cMY  ChildishGambinoVEVO           10  2018-05-06T04:00:07.000Z   
FlsCjmMhFmw    YouTube Spotlight           24  2017-12-06T17:58:51.000Z   
ffxKSjUwKdU     ArianaGrandeVevo           10  2018-04-20T04:00:03.000Z   
zEf423kYfqk           BeckyGVEVO           10  2018-04-20T10:40:51.000Z   
7C2z4GqqS5E              ibighit           10  2018-05-18T09:00:02.000Z   

                                                          tags      views  \
video_id                                                                    
VYOjWnS4cMY  Childish Gambino|"Rap"|"This Is America"|"mcDJ...  225211923   
FlsCjmMhFmw  Rewind|"Rewind 2017"|"youtube rewind 2017"|"#Y...  149376127   
ffxKSjUwKdU  Ariana|"Grande"|"No"|"Tears"|"Left"|"To"|"Cry"...  148689896   
zEf423kYfqk  Becky G|"Natti Natasha"|"Natti Natasha Music"|...  139334502   
7C2z4GqqS5E  BIGHIT|"빅히트"|"방탄소년단"|"BTS"|"BANGTAN"|"방탄"|"FAK...  123010920   

               likes  dislikes  comment_count  \
video_id                                        
VYOjWnS4cMY  5023450    343541         517232   
FlsCjmMhFmw  3093544   1643059         827755   
ffxKSjUwKdU  3094021    129502         242039   
zEf423kYfqk  1425496    119798          83941   
7C2z4GqqS5E  5613827    206892        1228655   

                                             thumbnail_link  \
video_id                                                      
VYOjWnS4cMY  https://i.ytimg.com/vi/VYOjWnS4cMY/default.jpg   
FlsCjmMhFmw  https://i.ytimg.com/vi/FlsCjmMhFmw/default.jpg   
ffxKSjUwKdU  https://i.ytimg.com/vi/ffxKSjUwKdU/default.jpg   
zEf423kYfqk  https://i.ytimg.com/vi/zEf423kYfqk/default.jpg   
7C2z4GqqS5E  https://i.ytimg.com/vi/7C2z4GqqS5E/default.jpg   

             comments_disabled  ratings_disabled  video_error_or_removed  \
video_id                                                                   
VYOjWnS4cMY              False             False                   False   
FlsCjmMhFmw              False             False                   False   
ffxKSjUwKdU              False             False                   False   
zEf423kYfqk              False             False                   False   
7C2z4GqqS5E              False             False                   False   

                                                   description  
video_id                                                        
VYOjWnS4cMY  “This is America” by Childish Gambino http://s...  
FlsCjmMhFmw  YouTube Rewind 2017. Celebrating the videos, p...  
ffxKSjUwKdU  No Tears Left To Cry (Official Video)\nSong Av...  
zEf423kYfqk  Becky G & Natti Natasha – “Sin Pijama”\n\nNUEV...  
7C2z4GqqS5E  BTS (방탄소년단) 'FAKE LOVE' Official MVDirector : ...